File: Apps4Bro.iOS/Networks/AdMobNetwork.cs
#if USE_ADMOB && !OLD_ADMOB
using System;
using Foundation;
using Google.MobileAds;

namespace Apps4Bro.Networks
{
    internal class AdMobNetwork : FullScreenContentDelegate, AdNetworkHandler
    {
        private readonly AdManager m_adManager;
        private string m_id;
        private InterstitialAd m_gadInterstitial;
        private object m_data;
        private bool m_nonPersonalizedAds;

        public string Network
        {
            get { return "AdMob"; }
        }


        public AdMobNetwork(AdManager manager)
        {
            m_adManager = manager;
        }

        public bool Show(object data)
        {
            m_data = data;

            Hide();

            Request request = Request.GetDefaultRequest();
            if (m_nonPersonalizedAds || Apps4BroSDK.UseNonPersonalizedAds)
            {
                var extras = new Extras();
                extras.AdditionalParameters = NSDictionary.FromObjectAndKey(new NSString("1"), new NSString("npa"));
                request.RegisterAdNetworkExtras(extras);

            }
            InterstitialAd.Load(m_id, request, delegate (InterstitialAd interstitialAd, NSError error)
            {
                if (error != null)
                    m_adManager.AdError(m_data, "Error " + error.Code + ": " + error.Description);
                else
                {
                    m_gadInterstitial = interstitialAd;
                    m_gadInterstitial.Delegate = this;

                    m_adManager.AdLoaded(m_data);
                }

            });

            return true;
        }

        public void Hide()
        {
            if (m_gadInterstitial != null)
            {
                m_gadInterstitial.Delegate = null;
                m_gadInterstitial.Dispose();
                m_gadInterstitial = null;
            }
        }

        public void Display()
        {
            m_gadInterstitial.Present(UIKit.UIApplication.SharedApplication.Windows[0].RootViewController);
            // Nomad: I'm changing this one to use topmost view controller to avoid skipping ads with settings window, etc.

            //m_gadInterstitial.Present((UIKit.UIViewController)m_adManager.Context);
        }

        public void SetId(string id)
        {
            string[] split = id.Split(';');
            m_id = split[0];
            m_nonPersonalizedAds = split.Length > 1 && split[1].ToLower() == "true";
        }

        public override void DidPresentFullScreenContent(FullScreenPresentingAd ad)
        {
            m_adManager.AdShown(m_data, ad);
        }

        public override void DidFailToPresentFullScreenContent(FullScreenPresentingAd ad, NSError error)
        {
            m_adManager.AdError(m_data, "Error " + error.Code + ": " + error.Description);
        }

        public override void DidRecordClick(FullScreenPresentingAd ad)
        {
            m_adManager.AdClicked(m_data);
        }
    }
}
#endif

File: Apps4Bro.Swift/Sources/Apps4Bro/Networks/AdMobNetwork.swift
import Foundation
import UIKit
import GoogleMobileAds

// Port of AdMobNetwork.cs. C# used the legacy `InterstitialAd.Load(unit, request, ...)`
// API + `FullScreenContentDelegate`. Modern GoogleMobileAds Swift uses
// `InterstitialAd.load(with:request:completionHandler:)` + `FullScreenContentDelegate`.
final class AdMobNetwork: NSObject, AdNetworkHandler, GADFullScreenContentDelegate {
    private weak var manager: AdManager?
    private var unitId: String = ""
    private var interstitial: GADInterstitialAd?
    private var data: Any?
    private var nonPersonalizedAds: Bool = false

    let network: String = "AdMob"

    init(manager: AdManager) {
        self.manager = manager
    }

    func setId(_ id: String) {
        // C# split "<unit>;<bool>" — second part toggles NPA per-zone.
        let parts = id.split(separator: ";").map(String.init)
        unitId = parts.first ?? id
        nonPersonalizedAds = parts.count > 1 && parts[1].lowercased() == "true"
    }

    func show(data: Any) -> Bool {
        self.data = data
        hide()

        let request = GADRequest()
        if nonPersonalizedAds || Apps4BroSDK.useNonPersonalizedAds {
            let extras = GADExtras()
            extras.additionalParameters = ["npa": "1"]
            request.register(extras)
        }

        GADInterstitialAd.load(withAdUnitID: unitId, request: request) { [weak self] ad, error in
            guard let self = self else { return }
            if let error = error {
                let ns = error as NSError
                self.manager?.adError(data: self.data, error: "Error \(ns.code): \(ns.localizedDescription)")
                return
            }
            self.interstitial = ad
            self.interstitial?.fullScreenContentDelegate = self
            self.manager?.adLoaded(data: self.data)
        }
        return true
    }

    func hide() {
        interstitial?.fullScreenContentDelegate = nil
        interstitial = nil
    }

    func display() {
        // C# comment: present from the topmost view controller, not the original context,
        // to avoid skipping ads while settings windows are open.
        guard let root = AdMobNetwork.topViewController() else { return }
        interstitial?.present(fromRootViewController: root)
    }

    // MARK: - FullScreenContentDelegate

    func ad(_ ad: GADFullScreenPresentingAd, didFailToPresentFullScreenContentWithError error: Error) {
        let ns = error as NSError
        manager?.adError(data: data, error: "Error \(ns.code): \(ns.localizedDescription)")
    }

    func adDidRecordClick(_ ad: GADFullScreenPresentingAd) {
        manager?.adClicked(data: data)
    }

    func adDidRecordImpression(_ ad: GADFullScreenPresentingAd) {
        manager?.adShown(data: data, view: ad)
    }

    // MARK: - Helpers

    static func topViewController(_ base: UIViewController? = nil) -> UIViewController? {
        let root = base ?? UIApplication.shared.connectedScenes
            .compactMap { $0 as? UIWindowScene }
            .flatMap { $0.windows }
            .first(where: { $0.isKeyWindow })?
            .rootViewController
        if let nav = root as? UINavigationController { return topViewController(nav.visibleViewController) }
        if let tab = root as? UITabBarController { return topViewController(tab.selectedViewController) }
        if let presented = root?.presentedViewController { return topViewController(presented) }
        return root
    }
}
